File: src/infinicore/ops/gemm/gemm.cc
#include "infinicore/ops/gemm.hpp"

namespace infinicore::op {

common::OpDispatcher<Gemm::schema> &Gemm::dispatcher() {
    static common::OpDispatcher<Gemm::schema> dispatcher_;
    return dispatcher_;
};

void Gemm::execute(Tensor c, Tensor a, Tensor b, float alpha, float beta) {
    dispatcher().lookup(context::getDevice().getType())(c, a, b, alpha, beta);
}

Tensor gemm(Tensor a, Tensor b, float alpha, float beta) {
    Shape shape = a->shape();
    Size size = a->ndim();
    shape[size - 1] = b->size(size - 1);
    auto c = Tensor::empty(shape, a->dtype(), a->device());
    gemm_(c, a, b, alpha, beta);
    return c;
}

void gemm_(Tensor c, Tensor a, Tensor b, float alpha, float beta) {
    Gemm::execute(c, a, b, alpha, beta);
}

} // namespace infinicore::op

File: python/infinicore/ops/gemm.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def gemm(input, other, alpha=1.0, beta=0.0, *, out=None):
    if out is None:
        return Tensor(_infinicore.gemm(input._underlying, other._underlying, alpha, beta))

    _infinicore.gemm_(out._underlying, input._underlying, other._underlying, alpha, beta)
